### ISSUE TITLE
Define softmax semantics and implement a lowering to the correct sequence of named ops

### ISSUE THREAD

**Issue by kurapov-peter** [intel/graph-compiler#10]
For this, we need to lower it from multiple models and multiple frameworks and see "what they do". At least two potential options: `exp(arg)` and `exp(arg-mean)`, might be more. We need to pick one, provide reasoning, and then implement a lowering upstream.

**Comment by rengolin** [comment]
I'd start by looking what TensorFlow, PyTorch, ONNX, OpenVino, oneDNN graph do currently. Look at their definitions, if possible lower a model on their graph description to MLIR and see what comes, etc.

Once we know what people will use it for, we encode that as _"semantics"_ and lower it to that sequence of named ops. 

We may need to create a new pass in addition to the legalizer. Something like `linalg-decomposition`.

**Comment by kurapov-peter** [comment]
Sounds good, let's start by collecting data from frameworks.

**Comment by chudur-budur** [comment]
> I'd start by looking what TensorFlow, PyTorch, ONNX, OpenVino, oneDNN graph do currently. Look at their definitions, if possible lower a model on their graph description to MLIR and see what comes, etc.
> 
> Once we know what people will use it for, we encode that as _"semantics"_ and lower it to that sequence of named ops.
> 
> We may need to create a new pass in addition to the legalizer. Something like `linalg-decomposition`.

I tried with TensorFlow, they have a function that spits out MLIR from a function:
```python
import tensorflow as tf
from tensorflow.python.ops.gen_math_ops import exp as expf

expf = tf.function(expf)

cfunc = expf.get_concrete_function(tf.TensorSpec(None, tf.dtypes.float32))

mlir_str = tf.mlir.experimental.convert_function(cfunc, pass_pipeline="builtin.module(any(cse,canonicalize),convert-func-to-llvm{use-bare-ptr-memref-call-conv=1})")

print(mlir_str)
```

and I am getting this:

```llvm
module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, producer = 1766 : i32}} {
  func.func @__inference_exp_5(%arg0: tensor<*xf32> {tf._user_specified_name = "x"}) -> tensor<*xf32> attributes {allow_soft_placement = false, tf.entry_function = {control_outputs = "", inputs = "x", outputs = "identity_RetVal"}} {
    %0 = "tf.Exp"(%arg0) {device = ""} : (tensor<*xf32>) -> tensor<*xf32>
    %1 = "tf.Identity"(%0) {device = ""} : (tensor<*xf32>) -> tensor<*xf32>
    return %1 : tensor<*xf32>
  }
}
```

I am looking into how to get MLIR from `tf.Exp`.

**Comment by kurapov-peter** [comment]
Does their `tf-standard-pipeline` produce the same output? I wonder what is it exactly, too.

**Comment by chudur-budur** [comment]
> Does their `tf-standard-pipeline` produce the same output? I wonder what is it exactly, too.

I made a mistake, you were right, the output above was with `tf-standard-pipeline`, now it's fixed. 

The `tf.Exp` is an `exp()` based on `tf` dialect. https://www.tensorflow.org/mlir/tf_ops#tfexp_tfexpop

**Comment by chudur-budur** [comment]
Most likely `tf.Exp` is a "legacy call" OP defined here: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td;drc=ff9a2058aae80dd43b722caad7db88d631ffa7ea;l=443

**Comment by kurapov-peter** [comment]
Since TF created mlir as a part of the framework I'd assume they have some lowering in mind? Is it IREE?

**Comment by chudur-budur** [comment]
For the tensforlow part, I need tools like `tf-opt` and `tf-mlir-translate`. To do this I need to compile tensorflow from source. 

But I am having these issues:

https://github.com/tensorflow/tensorflow/issues/68247
https://github.com/tensorflow/tensorflow/issues/68246

Any idea?

**Comment by rengolin** [comment]
When we did the same thing over a year ago, we also had to build TF from source, which was painful. I created a [branch](https://github.com/plaidml/mlir-generator/blob/build-tf/scripts/build-tensorflow.sh) almost 2 years ago (so YMMV) to automate that.

However, I do recommend you use IREE directly (from pip install) and go from there. IREE is much easier to use.

**Comment by chudur-budur** [comment]
> When we did the same thing over a year ago, we also had to build TF from source, which was painful. I created a [branch](https://github.com/plaidml/mlir-generator/blob/build-tf/scripts/build-tensorflow.sh) almost 2 years ago (so YMMV) to automate that.

Thanks. But the script in that branch doesn't compile tf-apis. I managed to compile TF from source, but couldn't compile the tf-api (`tf-opt`, `tf-mlir-translate`) to a functional binary. I can compile but getting linker problems.

https://github.com/tensorflow/tensorflow/issues/69367

> However, I do recommend you use IREE directly (from pip install) and go from there. IREE is much easier to use.

IREE doesn't understand `tf_executor` dialect. Do you have any idea how do I register it to IREE?

**Comment by rengolin** [comment]
IIRC, you need to use `iree-import-tf` (https://iree.dev/guides/ml-frameworks/tensorflow/) and pull the IR from a saved model.

If trying TF directly, there is a `tf.mlir.experimental` (https://www.tensorflow.org/api_docs/python/tf/mlir/experimental/convert_graph_def) section that may help.

Once you have HLO, you should be able to use `iree-opt` to lower to Linalg (https://groups.google.com/g/iree-discuss/c/u2zqHKVqPTc/m/UCGeYoauGwAJ).

**Comment by kurapov-peter** [comment]
# PyTorch to torch-mlir summary

TL;DR the resulting linalg does `tmp=exp(arg - max(arg)); tmp / sum(tmp)`.

Input model:
```Python
class SoftMax(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.softmax = torch.nn.Softmax()

    def forward(self, x):
        return self.softmax(x)
```

The forward lowers to (1d case shown for simplicity, the IR is very similar for higher dims):
```MLIR
module attributes {torch.debug_module_name = "SoftMax"} {
  func.func @SoftMax(%arg0: !torch.vtensor<[128],f32>) -> !torch.vtensor<[128],f32> {
    %none = torch.constant.none
    %0 = torch.aten.clone %arg0, %none : !torch.vtensor<[128],f32>, !torch.none -> !torch.vtensor<[128],f32>
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %1 = torch.aten._softmax %0, %int0, %false : !torch.vtensor<[128],f32>, !torch.int, !torch.bool -> !torch.vtensor<[128],f32>
    return %1 : !torch.vtensor<[128],f32>
  }
}
```

Linalg:
```MLIR
#map = affine_map<(d0) -> (d0)>
#map1 = affine_map<(d0) -> ()>
#map2 = affine_map<(d0) -> (0)>
module attributes {torch.debug_module_name = "SoftMax"} {
  ml_program.global private mutable @global_seed(dense<0> : tensor<i64>) : tensor<i64>
  func.func @SoftMax(%arg0: tensor<128xf32>) -> tensor<128xf32> {
    %c0_i64 = arith.constant 0 : i64
    %cst = arith.constant 0xFF800000 : f32
    %cst_0 = arith.constant 0.000000e+00 : f32
    %0 = tensor.empty() : tensor<i64>
    %1 = linalg.fill ins(%c0_i64 : i64) outs(%0 : tensor<i64>) -> tensor<i64>
    %2 = tensor.empty() : tensor<f32>
    %3 = linalg.fill ins(%cst : f32) outs(%2 : tensor<f32>) -> tensor<f32>
    %4:2 = linalg.generic {indexing_maps = [#map, #map1, #map1], iterator_types = ["reduction"]} ins(%arg0 : tensor<128xf32>) outs(%3, %1 : tensor<f32>, tensor<i64>) {
    ^bb0(%in: f32, %out: f32, %out_1: i64):
      %12 = linalg.index 0 : index
      %13 = arith.index_cast %12 : index to i64
      %14 = arith.maximumf %in, %out : f32
      %15 = arith.cmpf ogt, %in, %out : f32
      %16 = arith.select %15, %13, %out_1 : i64
      linalg.yield %14, %16 : f32, i64
    } -> (tensor<f32>, tensor<i64>)
    %expanded = tensor.expand_shape %4#0 [] : tensor<f32> into tensor<1xf32>
    %5 = tensor.empty() : tensor<128xf32>
    %6 = linalg.generic {indexing_maps = [#map, #map2, #map], iterator_types = ["parallel"]} ins(%arg0, %expanded : tensor<128xf32>, tensor<1xf32>) outs(%5 : tensor<128xf32>) {
    ^bb0(%in: f32, %in_1: f32, %out: f32):
      %12 = arith.subf %in, %in_1 : f32
      linalg.yield %12 : f32
    } -> tensor<128xf32>
    %7 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%6 : tensor<128xf32>) outs(%5 : tensor<128xf32>) {
    ^bb0(%in: f32, %out: f32):
      %12 = math.exp %in : f32
      linalg.yield %12 : f32
    } -> tensor<128xf32>
    %8 = tensor.empty() : tensor<1xf32>
    %9 = linalg.fill ins(%cst_0 : f32) outs(%8 : tensor<1xf32>) -> tensor<1xf32>
    %10 = linalg.generic {indexing_maps = [#map, #map2], iterator_types = ["reduction"]} ins(%7 : tensor<128xf32>) outs(%9 : tensor<1xf32>) {
    ^bb0(%in: f32, %out: f32):
      %12 = arith.addf %in, %out : f32
      linalg.yield %12 : f32
    } -> tensor<1xf32>
    %11 = linalg.generic {indexing_maps = [#map, #map2, #map], iterator_types = ["parallel"]} ins(%7, %10 : tensor<128xf32>, tensor<1xf32>) outs(%5 : tensor<128xf32>) {
    ^bb0(%in: f32, %in_1: f32, %out: f32):
      %12 = arith.divf %in, %in_1 : f32
      linalg.yield %12 : f32
    } -> tensor<128xf32>
    return %11 : tensor<128xf32>
  }
}
```

**Comment by kurapov-peter** [comment]
OpenVINO & OneDNN Graph have similar implementations, following `tmp=exp(arg - max(arg)); tmp / sum(tmp)`.
We'll now look into generating a TF example for the complete picture.

**Comment by kurapov-peter** [comment]
# TensorFlow to IREE summary

TL;DR the resulting linalg also does tmp=exp(arg - max(arg)); tmp / sum(tmp).

Input model:
```Python
class Softmax(tf.keras.Model):
    def __init__(self):
        super(Softmax, self).__init__()

    def call(self, inputs, training=False):
      return tf.nn.softmax(inputs)
```

Forward lowers to:
```MLIR
module {
  util.func public @session_initializer() {
    util.return
  }
  util.func public @serving_default(%arg0: tensor<6xf32>) -> tensor<6xf32> {
    %0 = stablehlo.constant dense<-0.000000e+00> : tensor<f32>
    %1 = stablehlo.constant dense<0xFF800000> : tensor<f32>
    %2 = stablehlo.reduce(%arg0 init: %1) applies stablehlo.maximum across dimensions = [0] : (tensor<6xf32>, tensor<f32>) -> tensor<f32>
    %3 = stablehlo.reshape %2 : (tensor<f32>) -> tensor<1xf32>
    %4 = stablehlo.broadcast_in_dim %3, dims = [0] : (tensor<1xf32>) -> tensor<6xf32>
    %5 = stablehlo.subtract %arg0, %4 : tensor<6xf32>
    %6 = stablehlo.exponential %5 : tensor<6xf32>
    %7 = stablehlo.reduce(%6 init: %0) applies stablehlo.add across dimensions = [0] : (tensor<6xf32>, tensor<f32>) -> tensor<f32>
    %8 = stablehlo.reshape %7 : (tensor<f32>) -> tensor<1xf32>
    %9 = stablehlo.broadcast_in_dim %8, dims = [0] : (tensor<1xf32>) -> tensor<6xf32>
    %10 = stablehlo.divide %6, %9 : tensor<6xf32>
    util.return %10 : tensor<6xf32>
  }
}
```

Resulting linalg:
```MLIR
// -----// IR Dump After ConvertStableHloToIreeInputDialects (iree-stablehlo-to-iree-input) //----- //
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", native_vector_size = 16 : i64, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
#map = affine_map<(d0) -> (d0)>
#map1 = affine_map<(d0) -> ()>
#map2 = affine_map<(d0) -> (0)>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", [#executable_target_embedded_elf_x86_64_]>
module attributes {hal.device.targets = [#device_target_llvm_cpu]} {
  func.func @session_initializer() {
    return
  }
  func.func @serving_default(%arg0: tensor<6xf32>) -> tensor<6xf32> {
    %cst = arith.constant -0.000000e+00 : f32
    %cst_0 = arith.constant 0xFF800000 : f32
    %0 = tensor.empty() : tensor<f32>
    %1 = linalg.fill ins(%cst_0 : f32) outs(%0 : tensor<f32>) -> tensor<f32>
    %2 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["reduction"]} ins(%arg0 : tensor<6xf32>) outs(%1 : tensor<f32>) {
    ^bb0(%in: f32, %out: f32):
      %16 = arith.maximumf %out, %in : f32
      linalg.yield %16 : f32
    } -> tensor<f32>
    %expanded = tensor.expand_shape %2 [] : tensor<f32> into tensor<1xf32>
    %3 = tensor.empty() : tensor<6xf32>
    %4 = linalg.generic {indexing_maps = [#map2, #map], iterator_types = ["parallel"]} ins(%expanded : tensor<1xf32>) outs(%3 : tensor<6xf32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<6xf32>
    %5 = tensor.empty() : tensor<6xf32>
    %6 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel"]} ins(%arg0, %4 : tensor<6xf32>, tensor<6xf32>) outs(%5 : tensor<6xf32>) {
    ^bb0(%in: f32, %in_2: f32, %out: f32):
      %16 = arith.subf %in, %in_2 : f32
      linalg.yield %16 : f32
    } -> tensor<6xf32>
    %7 = tensor.empty() : tensor<6xf32>
    %8 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%6 : tensor<6xf32>) outs(%7 : tensor<6xf32>) {
    ^bb0(%in: f32, %out: f32):
      %16 = math.exp %in : f32
      linalg.yield %16 : f32
    } -> tensor<6xf32>
    %9 = tensor.empty() : tensor<f32>
    %10 = linalg.fill ins(%cst : f32) outs(%9 : tensor<f32>) -> tensor<f32>
    %11 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["reduction"]} ins(%8 : tensor<6xf32>) outs(%10 : tensor<f32>) {
    ^bb0(%in: f32, %out: f32):
      %16 = arith.addf %out, %in : f32
      linalg.yield %16 : f32
    } -> tensor<f32>
    %expanded_1 = tensor.expand_shape %11 [] : tensor<f32> into tensor<1xf32>
    %12 = tensor.empty() : tensor<6xf32>
    %13 = linalg.generic {indexing_maps = [#map2, #map], iterator_types = ["parallel"]} ins(%expanded_1 : tensor<1xf32>) outs(%12 : tensor<6xf32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<6xf32>
    %14 = tensor.empty() : tensor<6xf32>
    %15 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel"]} ins(%8, %13 : tensor<6xf32>, tensor<6xf32>) outs(%14 : tensor<6xf32>) {
    ^bb0(%in: f32, %in_2: f32, %out: f32):
      %16 = arith.divf %in, %in_2 : f32
      linalg.yield %16 : f32
    } -> tensor<6xf32>
    return %15 : tensor<6xf32>
  }
}
```

**Comment by kurapov-peter** [comment]
All the reviewed frameworks/paths follow the same semantics. We can proceed with the lowering implementation.

**Comment by chudur-budur** [comment]
An example upstream linalg op implementation
https://github.com/llvm/llvm-project/pull/90236

**Comment by rengolin** [comment]
Yes, upstream must have a lowering (currently generics) to get to run at all. The point here is to lower to named ops instead.

And if we do that, the questions that need answering are:
* what is the most appropriate lowering (ops, casts, shapes)? 
* does the current generic lowering uses that?
* can we lower to the same semantic for generic and named ops?
* do we always lower to named ops, and then if the user wants generics, we generalize them? Or do we parametrize the lowering pass?

**Comment by rengolin** [comment]
https://github.com/llvm/llvm-project/pull/96090

**Comment by kurapov-peter** [comment]
Update: the first implementation attempt is here https://github.com/llvm/llvm-project/compare/main...kurapov-peter:llvm-project:softmax?expand=1. Now looking into how to merge it with the `AggregatedOpInterface`. Current transform decomposition skips to generics. The idea now (WIP) is to change the interface implementation and modify the existing showcasing transform test to additionally call `transform.structured.generalize` after decomposition to named ops.
